thunder/tests/distributed/test_ddp.py

```diff
@@ -1021,6 +1021,51 @@ def fwd_loss(m, x):
         with thunder.ThunderModule.no_sync(model):
             fwd_loss(model, x)
 
+    @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="Requires 2 devices")
+    def test_fsdp_weight_sharing(self):
+        # This test is to verify that weight sharing works with fsdp.
+        # NOTE: Currently we end up creating 2 copies of shared weight during execution.
+        #       This should be fixed and we should update this test to check for that.
+        device = torch.device("cuda", self.rank)
+        torch.set_default_device(device)
+
+        class Model(torch.nn.Module):
+            def __init__(self) -> None:
+                super().__init__()
+                self.fc1 = torch.nn.Linear(16, 16, bias=False)
+                self.fc2 = torch.nn.Linear(16, 16, bias=False)
+
+            def forward(self, x):
+                return self.fc1(x) + self.fc2(x)
+
+        # Check `jit(fsdp(model))` works
+        with device:
+            model = Model()
+
+        model.fc1.weight = model.fc2.weight
+
+        model = thunder.jit(thunder.distributed.fsdp(model), executors=["torch"])
+
+        x = torch.randn(4, 16)
+        output = model(x)
+
+        expected_shape = (4, 16)
+        assert output.shape == expected_shape, f"{output.shape=} - {expected_shape=}"
+
+        # Check `fsdp(jit(model))` works
+        with device:
+            model = Model()
+
+        model.fc1.weight = model.fc2.weight
+
+        model = thunder.distributed.fsdp(thunder.jit(model, executors=["torch"]))
+
+        x = torch.randn(4, 16)
+        output = model(x)
+
+        expected_shape = (4, 16)
+        assert output.shape == expected_shape, f"{output.shape=} - {expected_shape=}"
```

Comment:
I think things get sharded twice here (but with separate results) and we should avoid this, right?

Reply:
For `fsdp(jit(model))` case, we do shard the parameter twice but we shard the same original parameter from user passed module (i.e. we don't reshard the already sharded parameter). So, both copies actually hold the correct data. Ideally we should have only one copy. I am planning to look into this in a separate PR for both `fsdp(jit(model))` and `jit(fsdp(model))`. 

More details at https://github.com/Lightning-AI/lightning-thunder/issues/257#issuecomment-2133070264

We can verify this with the below snippet.

```python
import os
import torch
import torch.distributed as tdist
import thunder
import thunder.distributed

if __name__ == "__main__":
    tdist.init_process_group(backend="nccl")
    LOCAL_RANK = int(os.environ["LOCAL_RANK"])
    device = torch.device("cuda", LOCAL_RANK)
    torch.set_default_device(device)

    class Model(torch.nn.Module):
        def __init__(self) -> None:
            super().__init__()
            self.fc1 = torch.nn.Linear(16, 16, bias=False)
            self.fc2 = torch.nn.Linear(16, 16, bias=False)

        def forward(self, x):
            return self.fc1(x) + self.fc2(x)

    with device:
        model = Model()
        input_t = torch.randn(4, 16)

    model.fc1.weight = model.fc2.weight
    # if LOCAL_RANK == 0:
    #     print(model)
    # model = thunder.distributed.fsdp(model)
    # model = thunder.jit(model, executors=["torch"])

    model = thunder.jit(model, executors=["torch"])
    model = thunder.distributed.fsdp(model)

    # Verify that tensor data is same in both weights.
    torch.testing.assert_close(model._overrides["fc1.weight"], model._overrides["fc2.weight"])

    out = model(input_t)

    if LOCAL_RANK == 0:
        trace = thunder.last_traces(model)[-1]
        with open("weight_sharing_trace.py", 'w') as f:
            f.write(str(trace))

```

Reply:
Seems good! Thank you!

Reply:
Gradients for `fsdp(jit(...))` are incorrect, I am planning to fix `fsdp(jit(...))` in a seperate PR in a day or so.

Reply:
That's the "should avoid this", probably, as we would want the shared param to get the sum of the gradients.